### ISSUE TITLE
Declare parent host in /etc/hosts

### ISSUE THREAD

**Issue by eirslett** [CiscoCloud/microservices-infrastructure#600]
We want our Docker containers to talk to their parents/worker nodes, e.g. for collectd, logstash, consul k/v.
It would be nice to have a "mesos-host" entry in /etc/hosts in every worker node, so you can access consul at https://mesos-host:8500 to be sure that you get the local consul agent and don't do any network hops. (Same for logstash and collectd, or in general)
Does that sound like a good solution to the problem of contacting hosts from Docker containers?

**Comment by stevendborrelli** [comment]
This is a good idea, I wonder what other groups do to send the host IP into docker containers. 

It seems env variables might be another way to do it? maybe DOCKER_HOST or something like that? 

http://stackoverflow.com/questions/22944631/how-to-get-the-ip-address-of-the-docker-host-from-inside-a-docker-container

**Comment by eirslett** [comment]
Environment variables is an alternative. It looks like most people use that in combination with heuristic shell scripts to determine the IP. This project is lucky because ansible knows all the IP addresses at the time it writes to /etc/hosts; so no heuristics needed.
The good thing about writing to /etc/hosts vs. setting an environment variable is that the first will be usable in a wider scope. E.g. from mesos tasks that are not Docker tasks (rkt maybe, or just simple apps running on the host itself) Also, there's no easy way I know of to pass additional environment variables to Docker via Marathon. (and must also be done for Chronos, and for Kubernetes, etc etc) With an entry in /etc/hosts that would come for free in every case.

**Comment by eirslett** [comment]
One could also use that hosts entry to refer to consul/logstash/collectd running on localhost etc.

**Comment by siddharthist** [comment]
Looks like [Marathon has the ability to pass in environment variables](https://mesosphere.github.io/marathon/docs/rest-api.html#post-v2-apps), so we could do that with Ansible templates. 

@eirslett Would writing to the host's /etc/hosts file change what hosts are available in child docker containers?

**Comment by eirslett** [comment]
Yes, the docker containers get the data from /etc/hosts. But I think that happens via dnsmasq, and /etc/hosts doesn't allow you to define wildcard entries either, which dnsmasq can. (So this change should be in the static dnsmasq configuration instead of /etc/hosts)

Docker containers resolve dns lookups through consul, which in turn resolves not-consul-domains using the default dns resolver, which in mantl's case is dnsmasq, which in turn looks at /etc/hosts in addition to its own rules.
So it's docker -> consul -> dnsmasq -> /etc/hosts
Or maybe it's the other way around, docker -> dnsmasq -> consul -> /etc/hosts. Either way, changing it in dnsmasq should do the job.
We should be able to add a wildcard dns "local-lb" with dnsmasq, and also a simple "worker" record, so you can route internal traffic to "http://my-service-a.local-lb/foo/bar" which will delegate to the load balancer (traefik) running on the local node. And you can set up logstash forwarding to the "worker" dns record. Then you can set LOGSTASH_ENDPOINT="tcp://worker:1337" as an environment variable. Or https://worker:8500, where consul is running, which is a great improvement over https://consul.service.consul:8500, because you're guaranteed to be talking to the local consul agent instead of a random one.

**Comment by siddharthist** [comment]
So if I understand correctly: we should change `/etc/dnsmasq.conf` to include 

 1. An entry for the local Consul entry, e.g. `CONSUL_ENDPOINT={{ private_ip4 }}:8500`
 2. An entry for the local logstash entry, e.g. `LOGSTASH_ENDPOINT={{ private_ip4 }}:1337`
 3. An entry for the parent in general, e.g. `DOCKER_HOST={{ private_ip4 }}`

Is that correct @eirslett?

**Comment by eirslett** [comment]
Uhm, something like that. I'm not sure why we need to override DOCKER_HOST? Isn't that done in calico already? Or defaulting to the unix socket? Maybe DOCKER_HOST is a bad example since it's an environment variable that already carries a different meaning.

http://stackoverflow.com/questions/22313142/wildcard-subdomains-with-dnsmasq

I'm not sure about the details of configuring dnsmasq, but it looks like we would need something like this: (just copy/pasting stuff from StackOverflow)

```
address=/worker/{{ public_ipv4 }}
address=/.local-lb/{{ public_ipv4 }}
```

And a distributive test that runs `dig` on "worker", "test-a.local-lb" and "test-b.local-lb" to verify that they resolve to public_ipv4.

**Comment by eirslett** [comment]
CONSUL_ENDPOINT and LOGSTASH_ENDPOINT would then not be dnsmasq configuration, but configuration that can be passed to applications. For example in marathon, if you want to start a new application that is supposed to connect to consul:
```json
{
  "env": {
    "CONSUL_ENDPOINT": "worker:8500"
  }
}
```

**Comment by siddharthist** [comment]
@eirslett Oh, I wasn't aware of `DOCKER_HOST` already being set, replace mentions with `DOCKER_PARENT` or similar. So in summary, we want to configure a DNS entry (probably via dnsmasq configuration) for `worker`, which points to the container's parent's IP address?

**Comment by eirslett** [comment]
Yup, exactly. And it has to support wildcards, so that the local load balancer/traefik can leverage virtualhosts to route traffic correctly.

**Comment by siddharthist** [comment]
I don't think I understand enough about DNS to actually move forward on this! I can add an entry in `/etc/hosts`, but I don't think that's what we're looking for. Anybody want to take this on?

**Comment by Zogg** [comment]
Maybe we could bind node's services on 0.0.0.0 interface, or on docker0's iface, instead of eth0's. This way processes from container could use docker0's ip to reach node's services. No need to mess with files in container.

**Comment by eirslett** [comment]
@Zogg it's not really about reaching the service on a network layer, it's about reaching a service at the HTTP application layer. Haproxy and Traefik typically route traffic based on subdomains; that's why we need a DNS record.
(You could in theory hack your applications, so that they override the HTTP Host header, then you wouldn't need a DNS record, but it's better not to ask business applications to take care of such things)
